src/Uno.Toolkit.UITest/Controls/SafeArea/Given_SafeArea.cs

```diff
@@ -223,36 +225,140 @@ public void When_Inside_Modal()
-			var flags = items
-				.Select((x, i) => new { Item = x, Flag = 1 << i })
-				.ToArray();
-			var max = (int)Math.Pow(2, items.Length);
-			for (int count = 0; count < max; count++)
-			{
-				yield return flags
-					.Where(x => ((count & x.Flag) == x.Flag))
-					.Select(x => x.Item)
-					.ToList();
-			}
-		}
-
-		private void WaitForIsChecked(QueryEx element, bool expected) =>
-			App.WaitForDependencyPropertyValue(element, "IsChecked", expected);
-	}
-}
+		}
+
+		[Test]
+		[TestCase("Soft", "Padding")]
+		[TestCase("Soft", "Margin")]
+		[TestCase("Hard", "Padding")]
+		[TestCase("Hard", "Margin")]
+		[ActivePlatforms(Platform.Android, Platform.iOS)]
+		public void When_SoftInput_Scroll_Into_View(string constraint, string mode)
+		{
+			const string TopTextBox = "TopTextBox";
+			const string BottomTextBox = "BottomTextBox";
+			const string SpacerBorder = "SpacerBorder";
+			var isSoft = string.Equals("Soft", constraint, StringComparison.OrdinalIgnoreCase);
+
+			NavigateToNestedSample("SafeArea_SoftInput_Scroll");
+
+			App.WaitForElement(SpacerBorder, timeoutMessage: $"Timed out waiting for element: {SpacerBorder}");
+
+			var keyboardRect = GetKeyboardRect();
+
+			App.FastTap($"{constraint}ConstraintMode");
+			App.FastTap($"{mode}Mode");
+
+			App.FastTap(TopTextBox);
+
+			App.Wait(TimeSpan.FromSeconds(1));
+
+			// Make sure the keyboard has opened and the focused TextBox is visible above the keyboard
+			Assert.Greater(keyboardRect.Height, 0);
+
+			if (isSoft)
+			{
+				var topRect = GetRectangle(TopTextBox);
+				var bottomRect = GetRectangle(BottomTextBox);
+
+				Assert.False(topRect.IsEmpty);
+				Assert.False(bottomRect.IsEmpty);
+
+				// The focused TextBox should be visible above the keyboard
+				Assert.AreEqual(bottomRect.Bottom, keyboardRect.Top);
+			}
+			else
+			{
+				var topRect = GetRectangle(TopTextBox);
+				var bottomRect = GetRectangle(BottomTextBox);
+
+				Assert.False(topRect.IsEmpty);
+
+				PlatformHelpers.On(
+					iOS: () => Assert.Greater(bottomRect.Top, keyboardRect.Top),
+					Android: () => Assert.True(bottomRect.IsEmpty)
+				);
+			}
+
+			App.DismissKeyboard();
+
+			App.FastTap(BottomTextBox);
+
+			App.Wait(TimeSpan.FromSeconds(1));
+
+			if (isSoft)
+			{
+				var topRect = GetRectangle(TopTextBox);
+				var bottomRect = GetRectangle(BottomTextBox);
+
+				Assert.False(topRect.IsEmpty);
+				Assert.False(bottomRect.IsEmpty);
+
+				Assert.Less(topRect.Bottom, keyboardRect.Top);
+				Assert.AreEqual(bottomRect.Bottom, keyboardRect.Top);
+			}
+			else
+			{
+				App.WaitForNoElement(TopTextBox, timeoutMessage: $"Timed out waiting for no element: {TopTextBox}");
+				var bottomRect = GetRectangle(BottomTextBox);
+
+				Assert.False(bottomRect.IsEmpty);
+				Assert.AreEqual(bottomRect.Bottom, keyboardRect.Top);
+			}
+
+			Rectangle GetKeyboardRect()
+			{
+				var initialBottom = App.GetPhysicalRect(BottomTextBox).Bottom;
+
+				App.FastTap($"SoftConstraintMode");

```

Comment:
not sure if i follow the logic, but shouldnt it be?:
```suggestion
				App.FastTap($"{constraint}ConstraintMode");
```

Reply:
Yeah this logic is just to force the keyboard to open so I can grab the change in position of the TextBox in order to get the height of the keyboard. I guess this would work for both constraint modes actually

Reply:
that could go in the comment